File: old/schedule_controller.rs
use crate::course::Course;
use crate::meeting::Meeting;
use crate::schedule::Schedule;
use crate::schedule_parameters::ScheduleParameters;
use crate::tools::{parse_days, parse_time, Days};
use futures::executor::block_on;
use itertools::Itertools;
use sqlx::SqlitePool;
use std::collections::HashMap;
use time::Time;

/// The `ScheduleController` type is an entry point for easy API interaction without any
/// manual calls to subclasses.
pub struct ScheduleController {
    pub schedules: Vec<Schedule>,
    course_list: Vec<String>,
    schedule_parameters: ScheduleParameters,
}

impl ScheduleController {
    pub fn new(schedule_parameters: ScheduleParameters, course_list: Vec<String>) -> Self {
        ScheduleController {
            schedules: Vec::new(),
            course_list,
            schedule_parameters,
        }
    }

    /// Generates all potential valid and non-overlapping course schedules.
    pub fn generate_schedules(&mut self) {
        let all_classes: Vec<Course> = match block_on(self.retrieve_all_classes()) {
            Ok(val) => val,
            Err(_) => panic!("Unable to generate schedules!"),
        };
        // Generate every possible combination of schedules (including overlapping)
        let mut combinations = all_classes.into_iter().combinations(self.course_list.len());

        // Filter schedules to ones that do not overlap
        for schedule in combinations {
            let s = Schedule::new(schedule, 0);
            if s.is_valid() {
                self.schedules.push(s);
            }
        }
    }

    // Private helper function for `generate_schedules`. Generates a list of all courses with correct metadata.
    async fn retrieve_all_classes(&self) -> Result<Vec<Course>, sqlx::Error> {
        let pool = SqlitePool::new("sqlite://src/classes.db").await?;
        let mut all_classes: Vec<Course> = Vec::new();

        for course in self.course_list.iter() {
            let mut cursor = sqlx::query!(
                "
                SELECT course.*
                FROM course
                WHERE course.course
                LIKE ?
                ",
                course
            )
            // .bind(course)
            .fetch_all(&pool)
            .await?;

            for row in cursor {
                let row_id: String = row.id.clone();
                println!("{}", row_id);

                // Generate the meetings array
                let mut course_meetings = sqlx::query!(
                    "
                    SELECT meeting.meeting_id, meeting.meeting_type, meeting.hours, meeting.days, meeting.location, meeting.instructor
                    FROM meeting
                    WHERE meeting.course_id = ?
                    ",
                    row_id.clone()
                )
                .fetch_all(&pool)
                .await?;
                let mut meetings: Vec<Meeting> = Vec::new();
                for meeting in course_meetings {
                    // Variables are defined here so they do not need to be recreated on the for loop below
                    let parsed_days: Vec<Days> = parse_days(&meeting.days.unwrap());
                    let meeting_id: String = meeting.meeting_id;
                    let meeting_type: String = meeting.meeting_type.unwrap();
                    let hours: [Time; 2] = parse_time(&meeting.hours.unwrap());
                    let location: String = meeting.location.unwrap();
                    let instructor: String = meeting.instructor.unwrap();
                    for individual_day in parsed_days {
                        let m: Meeting = Meeting {
                            meeting_id: meeting_id.clone(),
                            meeting_type: meeting_type.clone(),
                            hours,
                            day: individual_day,
                            location: location.clone(),
                            instructor: instructor.clone(),
                        };

                        meetings.push(m);
                    }
                }

                // Generate the footnotes array
                let mut course_footnotes = sqlx::query!(
                    "
                    SELECT footnote.code, footnote.text
                    FROM footnote
                    WHERE footnote.course_id = ?
                    ",
                    row_id.clone()
                )
                .fetch_all(&pool)
                .await?;
                let mut footnotes: HashMap<String, String> = HashMap::new();
                for footnote in course_footnotes {
                    footnotes.insert(footnote.code.unwrap(), footnote.text.unwrap());
                }
                let seats_open: i32 = row.seats_open.unwrap();
                let waitlist: bool = &seats_open == &0;
                let new_course = Course {
                    course_hours: row.course_hours.unwrap(),
                    course_title: row.course_title.unwrap(),
                    course: row.course.unwrap(),
                    description: row.description.unwrap(),
                    footnotes,
                    full_title: row.full_title.unwrap(),
                    general_text: row.general_text.unwrap(),
                    id: row.id,
                    meetings,
                    period: row.period.unwrap(),
                    prerequisite: row.prerequisite.unwrap(),
                    schedule_num: row.schedule_num.unwrap(),
                    seats_available: row.seats_available.unwrap(),
                    seats_open,
                    section: row.section.unwrap(),
                    session: row.session.unwrap(),
                    statement: row.statement.unwrap(),
                    units: row.units.unwrap(),
                    url: row.url.unwrap(),
                    waitlist,
                };
                all_classes.push(new_course);
            }
        }

        Ok(all_classes)
    }

    /// Returns the schedule with the highest fitness score.
    pub fn best_schedule(&self) -> Option<&Schedule> {
        match self.schedules.get(0) {
            Some(s) => return Some(s),
            None => None,
        }
    }
}

File: hourglass/schedule_controller.py
import itertools
import logging
from datetime import time
from typing import List

from sqlalchemy import create_engine, text

from hourglass.course import Course
from hourglass.datetime import DateTime, Day
from hourglass.meeting import Meeting
from hourglass.schedule import Schedule
from hourglass.schedule_parameters import ScheduleParameters


class ScheduleController:
    """
    ScheduleController is an abstraction for easy interaction with the database.

    In essence, ScheduleController provides an API for higher-level calls.

    Parameters
    ----------
    schedule_parameters : ScheduleParameters
        User-defined schedule parameters.
    course_list : List[str]
        List of strings which represent the courses to generate schedules with.
    """

    def __init__(
        self,
        schedule_parameters: ScheduleParameters = ScheduleParameters(),
        course_list: List[str] = [],
    ):
        self.schedule_parameters = schedule_parameters
        self.course_list = course_list
        self.schedules: List[Schedule] = []

    def generate_schedules(self):
        """Generate all potential, valid schedules."""
        all_classes: List[List[Course]] = []
        for course in self.course_list:
            c: List[Course] = self.generate_courses(course)
            all_classes.append(c)

        logging.debug(f"Found the following courses: {str(all_classes)}")

        for combination in itertools.product(*all_classes):
            s: Schedule = Schedule(combination)
            if s.is_valid():
                self.schedules.append(s)

    def generate_courses(self, course_string: str) -> List[Course]:
        """
        Generate Course objects for every course matching the given input.

        Parameters
        ----------
        course_string : str
            The name of the course to create a Course for.

        Returns
        -------
        List[Course]
            List of all courses matching the given input string.
        """
        all_courses = []
        engine = create_engine("sqlite:///classes.db")
        conn = engine.connect()
        query = text(
            """
            SELECT *
            FROM course
            WHERE course.course
            LIKE :course
            """
        )
        courses = conn.execute(query, course=course_string).fetchall()
        for course in courses:
            waitlist = course.seats_open == 0
            id = course.id
            # Meetings
            meetings = []
            query = text(
                """
                SELECT *
                FROM meeting
                WHERE meeting.course_id
                LIKE :course
                """
            )
            sql_meetings = conn.execute(query, course=id).fetchall()
            for meeting in sql_meetings:
                days: List[Day] = Day.parse_days(meeting.days)
                hours: List[time] = DateTime.parse_time(meeting.hours)
                for day in days:
                    dt: DateTime = DateTime(day=day, start=hours[0], end=hours[1])
                    m = Meeting(
                        date=dt,
                        instructor=meeting.instructor,
                        location=meeting.location,
                        meeting_id=meeting.meeting_id,
                        meeting_type=meeting.meeting_type,
                    )
                    meetings.append(m)
            # Handle meetings that overlap
            start_times = [meeting.date.start for meeting in meetings]
            if len(start_times) == len(set(start_times)):
                overlaps = False
            else:
                overlaps = True

            footnotes = {}
            query = text(
                """
                SELECT footnote.code, footnote.text
                FROM footnote
                WHERE footnote.course_id
                LIKE :course
                """
            )
            sql_footnotes = conn.execute(query, course=id).fetchall()
            for footnote in sql_footnotes:
                footnotes[footnote.code] = footnote.text

            c = Course(
                course_hours=course.course_hours,
                course_title=course.course_title,
                course=course.course,
                description=course.description,
                footnotes=footnotes,
                full_title=course.full_title,
                general_text=course.general_text,
                id=id,
                meetings=meetings,
                overlaps=overlaps,
                period=course.period,
                prerequisite=course.prerequisite,
                schedule_num=course.schedule_num,
                seats_available=course.seats_available,
                seats_open=course.seats_open,
                section=course.section,
                session=course.session,
                statement=course.statement,
                units=course.units,
                url=course.url,
                waitlist=waitlist,
            )
            all_courses.append(c)

        return all_courses

    # async def generate_courses(self, course_string: str):
    #     # TODO: figure this shit out with sqlalchemy
    #     courses = []
    #     async with aiosqlite.connect("sqlite://classes.db?mode=ro") as db:
    #         db.row_factory = aiosqlite.Row
    #         async with db.execute(
    #             """
    #         SELECT course.id
    #         FROM course
    #         WHERE course.course
    #         LIKE ?
    #         """,
    #             course_string,
    #         ) as cursor:
    #             async for row in cursor:
    #                 waitlist = row["seats_open"] == 0
    #                 id = row["id"]
    #                 meetings = await self.generate_meetings(id)
    #                 footnotes = await self.generate_footnotes(id)
    #                 c = Course(
    #                     course_hours=row["course_hours"],
    #                     course_title=row["course_title"],
    #                     course=row["course"],
    #                     description=row["description"],
    #                     footnotes=footnotes,
    #                     full_title=row["full_title"],
    #                     general_text=row["general_text"],
    #                     id=id,
    #                     meetings=meetings,
    #                     period=row["period"],
    #                     prerequisite=row["prerequisite"],
    #                     schedule_num=row["schedule_num"],
    #                     seats_available=row["seats_available"],
    #                     seats_open=row["seats_open"],
    #                     section=row["section"],
    #                     session=row["session"],
    #                     statement=row["statement"],
    #                     units=row["units"],
    #                     url=row["url"],
    #                     waitlist=waitlist,
    #                 )

    #                 courses.append(c)
    #     return courses

    # async def generate_meetings(self, course_id: str) -> List[Meeting]:
    #     meetings = []
    #     async with aiosqlite.connect("sqlite://classes.db?mode=ro") as db:
    #         db.row_factory = aiosqlite.Row
    #         async with db.execute(
    #             """
    #                 SELECT *
    #                 FROM meeting
    #                 WERE meeting.course_id
    #                 LIKE ?
    #                 """,
    #             course_id,
    #         ) as cursor:
    #             async for row in cursor:
    #                 days: List[Day] = Day.parse_days(row["days"])
    #                 hours: List[time] = DateTime.parse_time(row["hours"])
    #                 for day in days:
    #                     dt: DateTime = DateTime(day=day, start=hours[0], end=hours[1])
    #                     m = Meeting(
    #                         date=dt,
    #                         instructor=row["instructor"],
    #                         location=row["location"],
    #                         meeting_id=row["meeting_id"],
    #                         meeting_type=row["meeting_type"],
    #                     )
    #                     meetings.append(m)

    #     return meetings

    # async def generate_footnotes(self, course_id: str) -> Dict[str, str]:
    #     footnotes = {}
    #     async with aiosqlite.connect("sqlite://classes.db?mode=ro") as db:
    #         db.row_factory = aiosqlite.Row
    #         async with db.execute(
    #             """
    #             SELECT footnote.code, footnote.text
    #             FROM footnote
    #             WHERE footnote.course_id
    #             LIKE ?
    #                 """,
    #             course_id,
    #         ) as cursor:
    #             async for row in cursor:
    #                 footnotes[row["code"]] = row["text"]

    #     return footnotes

    def best_schedule(self) -> Schedule:
        [
            schedule.calculate_fitness(self.schedule_parameters)
            for schedule in self.schedules
        ]
        self.schedules.sort(reverse=True)
        return self.schedules[0]

    def __str__(self):
        output = ""
        for schedule in self.schedules:
            output += str(schedule) + "\n"
        return output
